File: iOS/WatBackend/WatBackend/RouteFinder.m

#import <Foundation/Foundation.h>
#import "RouteFinder.h"
#import "Path.h"
#import "Location.h"
#import "Map.h"

@interface LocationAndDistance : NSObject

@property Location* location;
@property double distance;

- (LocationAndDistance*) initWithLocation:(Location*) loc andDistance:(double)dist;

@end

@implementation LocationAndDistance

- (LocationAndDistance*) initWithLocation:(Location *)loc andDistance:(double)dist{
    _location = loc;
    _distance = dist;
    return self;
}

- (NSString*) description{
    return [NSString stringWithFormat:@"%0.1f:%@", _distance, _location];
}

@end

@implementation RouteFinder

- (RouteFinder*) initWithMap:(Map *)map{
    _map = map;
    return self;
}

- (Route*) findRouteFrom:(Building *)build1 To:(Building *)build2{
    double min_cost = DBL_MAX;
    Route *min_route = nil;
    Location *start_location = [build1 getMainFloor];
    
    for(Location *end_location in [build2 getAllFloors]){
        Route *cur_route = [self findRouteFromLoc:start_location To:end_location];
        double cur_cost = [cur_route getTotalCost];
        
        if(cur_cost < min_cost){
            min_cost = cur_cost;
            min_route = cur_route;
        }
    }
    
    return min_route;
}

- (Route*) findRouteFromLoc:(Location *)loc1 To:(Location *)loc2{
    NSArray *routelist = [self dijkstraFrom:loc1 To:loc2];
    
    Route *route = [[Route alloc] init];
    Location *cur = loc1;
    
    for(Location *loc in routelist){
        if([cur isEqual:loc]) continue;
        [route addStep:[[RouteStep alloc] initWithStart:cur withEnd:loc withPath:[_map retrievePathFrom:cur To:loc]]];
        cur = loc;
    }
    
    return route;
}

- (Route*) dijkstraFrom: (Location*) loc1 To:(Location*) loc2{
    NSMutableArray *queue = [[NSMutableArray alloc] init];
    NSMutableArray *visited = [[NSMutableArray alloc] init];
    
    NSMutableDictionary *distTable = [[NSMutableDictionary alloc] init];
    for(Location *loc in [_map locations]){
        distTable[loc] = [NSNumber numberWithDouble:DBL_MAX];
    }
    distTable[loc1] = [NSNumber numberWithDouble:0];
    
    [queue addObject:loc1];
    
    while([queue count] > 0){
        double min_cost = DBL_MAX;
        int min_index = 0;
        for(int i=0; i<[queue count]; i++){
            if([[distTable objectForKey:[queue objectAtIndex:i]] doubleValue] < min_cost){
                min_cost = [[distTable objectForKey:[queue objectAtIndex:i]] doubleValue];
                min_index = i;
            }
        }
        Location *cur = [queue objectAtIndex:min_index];
        [queue removeObjectAtIndex:min_index];
        
        NSArray *nextlist = [self getAdjacentLocations:cur];
        for(LocationAndDistance *next in nextlist){
            
            if(![visited containsObject:[next location]]){
                double alt_dist = [[distTable objectForKey:cur] doubleValue] + [next distance];
                double existing_dist = [[distTable objectForKey:[next location]] doubleValue];
                if(alt_dist < existing_dist){
                    distTable[[next location]] = [NSNumber numberWithDouble:alt_dist];
                    [queue addObject:[next location]];
                }
            }
            
        }
        
        [visited addObject:cur];
    }
    
    
    NSMutableArray *calcPath = [[NSMutableArray alloc] init];
    [calcPath addObject:loc2];
    
    Location *backtrack = loc2;
    while(![backtrack isEqual:loc1]){
        
        NSMutableArray *prevlist = [self getAdjacentLocations:backtrack];
        for(LocationAndDistance *prev in prevlist){
            if(fabs([prev distance] + [[distTable objectForKey:[prev location]] doubleValue] - [[distTable objectForKey:backtrack] doubleValue]) < 0.0001){
                [calcPath addObject:[prev location]];
                backtrack = [prev location];
            }
        }
        
    }
    
    calcPath = [[calcPath reverseObjectEnumerator] allObjects];
    
    return calcPath;
}

- (NSArray*) getAdjacentLocations: (Location*) loc{
    NSMutableArray *adjacents = [[NSMutableArray alloc] init];
    
    for(Path *path in [_map paths]){
        if([[path pointA] isEqual:loc]){
            double dist = [path getCost];
            [adjacents addObject:[[LocationAndDistance alloc] initWithLocation:[path pointB] andDistance:dist]];
        }
        if([[path pointB] isEqual:loc]){
            double dist = [path getCost];
            [adjacents addObject:[[LocationAndDistance alloc] initWithLocation:[path pointA] andDistance:dist]];
        }
    }
    
    return adjacents;
}

@end

File: android/WATisRain/RainBackend/src/main/java/com/lucky/watisrain/backend/RouteFinder.java
package com.lucky.watisrain.backend;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import com.lucky.watisrain.backend.data.Building;
import com.lucky.watisrain.backend.data.Location;
import com.lucky.watisrain.backend.data.Map;
import com.lucky.watisrain.backend.data.Path;
import com.lucky.watisrain.backend.data.Route;
import com.lucky.watisrain.backend.data.RouteStep;

/**
 * RouteFinder finds a Route given a Map. Details TBD.
 */
public class RouteFinder {
	
	Map map;
	
	/**
	 * Initialize with a map object
	 */
	public RouteFinder(Map map){
		this.map = map;
	}
	
	
	/**
	 * Find shortest Route object between two locations
	 */
	public Route findRoute(Location loc1, Location loc2){
		
		List<Location> routelist = dijkstra(loc1, loc2);
		
		Route route = new Route();
		Location cur = loc1;
		
		for(Location loc : routelist){
			if(cur.equals(loc)) continue;
			route.addStep(new RouteStep(cur, loc, map.retrievePath(cur, loc)));
			cur = loc;
		}
		
		return route;
	}
	
	
	/**
	 * Find the shortest route between main floor of build1 and ANY floor of build2
	 */
	public Route findRoute(Building build1, Building build2){
		
		double min_cost = Double.MAX_VALUE;
		Route min_route = null;
		Location start_location = build1.getMainFloor();
		
		for(Location end_location : build2.getAllFloors()){
			Route cur_route = findRoute(start_location, end_location);
			double cur_cost = cur_route.getTotalCost();
			
			if(cur_cost < min_cost){
				min_cost = cur_cost;
				min_route = cur_route;
			}
		}
		
		return min_route;
	}
	
	
	/**
	 * Return the shortest path between two locations, in the form of a List of Locations
	 */
	private List<Location> dijkstra(Location loc1, Location loc2){
		
		// Dijkstra's algorithm
		
		ArrayList<Location> queue = new ArrayList<>();
		ArrayList<Location> visited = new ArrayList<>();
		
		HashMap<Location, Double> distTable = new HashMap<>();
		for(Location loc : map.getLocations()){
			distTable.put(loc, Double.MAX_VALUE);
		}
		distTable.put(loc1, 0.0);
		
		queue.add(loc1);
		
		while(!queue.isEmpty()){
			
			// Extract min
			double min_cost = Double.MAX_VALUE;
			int min_index = 0;
			for(int i=0; i<queue.size(); i++){
				if(distTable.get(queue.get(i)) < min_cost){
					min_cost = distTable.get(queue.get(i));
					min_index = i;
				}
			}
			Location cur = queue.remove(min_index);
			
			List<LocationAndDistance> nextlist = getAdjacentLocations(cur);
			for(LocationAndDistance next : nextlist){
				
				if(!visited.contains(next.location)){
					double alt_dist = distTable.get(cur) + next.distance;
					double existing_dist = distTable.get(next.location);
					if(alt_dist < existing_dist){
						distTable.put(next.location, alt_dist);
						queue.add(next.location);
					}
				}
				
			}
			
			visited.add(cur);
		}
		
		
		ArrayList<Location> calcPath = new ArrayList<>();
		calcPath.add(loc2);
		
		// Construct the path backwards from destination
		Location backtrack = loc2;
		while(!backtrack.equals(loc1)){
			
			List<LocationAndDistance> prevlist = getAdjacentLocations(backtrack);
			for(LocationAndDistance prev : prevlist){
				if(Math.abs(prev.distance + distTable.get(prev.location)
						- distTable.get(backtrack)) < 0.0001){
					calcPath.add(prev.location);
					backtrack = prev.location;
				}
			}
			
		}
		
		Collections.reverse(calcPath);
		
		return calcPath;
	}
	
	
	/**
	 * List of Locations that are connected by a single path to a
	 * given Location
	 */
	private List<LocationAndDistance> getAdjacentLocations(Location loc){
		
		List<LocationAndDistance> adjacents = new ArrayList<>();
		
		for(Path path : map.getPaths()){
			if(path.getPointA().equals(loc)){
				double dist = path.getCost();
				adjacents.add(new LocationAndDistance(path.getPointB(), dist));
			}
			if(path.getPointB().equals(loc)){
				double dist = path.getCost();
				adjacents.add(new LocationAndDistance(path.getPointA(), dist));
			}
		}
		
		return adjacents;
	}
	
	
	/**
	 * Wrapper containing a location and distance
	 */
	class LocationAndDistance{
		
		Location location;
		double distance;
		
		public LocationAndDistance(Location loc, double dist){
			location = loc;
			distance = dist;
		}
		
		public String toString(){
			return distance + ":" + location;
		}
		
	}

}
